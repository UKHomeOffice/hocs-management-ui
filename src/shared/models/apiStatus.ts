export default interface ApiStatus {
    display: string;
    timeoutPeriod: number;
    type: string;
}
