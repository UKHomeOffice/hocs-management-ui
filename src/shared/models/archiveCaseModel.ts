export default interface ArchiveCaseModel {
    caseReference: string;
    deleted: boolean;
}
