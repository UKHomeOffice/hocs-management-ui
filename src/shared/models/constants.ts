export const ADD_CHILD_TOPIC_ERROR_DESCRIPTION = 'Something went wrong while adding the child topic. Please try again.';
export const ADD_STANDARD_LINE_ERROR_DESCRIPTION = 'Something went wrong while adding the standard line. Please try again.';
export const ADD_TEMPLATE_ERROR_DESCRIPTION = 'Something went wrong while adding the template. Please try again.';
export const ADD_UNIT_ERROR_DESCRIPTION = 'Something went wrong while adding the unit. Please try again.';
export const ADD_USER_ERROR_DESCRIPTION = 'Something went wrong while adding the following users. Please try again.';
export const ADD_USER_ERROR_TITLE = 'There was an error adding the users';
export const DUPLICATE_CHILD_TOPIC_DESCRIPTION = 'A child topic with that name already exists';
export const DUPLICATE_UNIT_DESCRIPTION = 'A unit with those details already exists';
export const EMPTY_SUBMIT_ERROR_DESCRIPTION = 'Please select some users before submitting.';
export const EMPTY_SUBMIT_ERROR_TITLE = 'No users selected';
export const EMPTY_SUBMIT_TEAM_ERROR_DESCRIPTION = 'Please select a team before submitting.';
export const EMPTY_SUBMIT_TEAM_ERROR_TITLE = 'No team selected';
export const EMPTY_SUBMIT_DRAFT_QA_TEAM_ERROR_DESCRIPTION = 'Select a team for Initial Draft and QA response stages.';
export const EMPTY_SUBMIT_PRIVATE_MINISTER_TEAM_ERROR_DESCRIPTION = 'Select a team for Private Office and Minister sign off stages.';
export const EMPTY_SUBMIT_TOPIC_ERROR_DESCRIPTION = 'Select a topic before submitting.';
export const EMPTY_SUBMIT_TOPIC_ERROR_TITLE = 'No topic selected';
export const GENERAL_ERROR_TITLE = 'Something went wrong';
export const LOAD_PARENT_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPIC_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_CASE_TYPES_DESCRIPTION = 'There was an error retrieving the case types.  Please try refreshing the page.';
export const LOAD_TEAM_ERROR_DESCRIPTION = 'There was an error retrieving the team.  Please try refreshing the page.';
export const LOAD_TEAM_MEMBERS_ERROR_DESCRIPTION = 'There was an error retrieving the teams members.  Please try refreshing the page.';
export const LOAD_TEAMS_ERROR_DESCRIPTION = 'There was an error retrieving the teams.  Please try refreshing the page.';
export const LOAD_USERS_ERROR_DESCRIPTION = 'There was an error retrieving the list of users.  Please try refreshing the page.';
export const REMOVE_FROM_TEAM_ALLOCATED_ERROR_DESCRIPTION = 'The user cannot be removed from the team as they have cases assigned';
export const REMOVE_FROM_TEAM_ERROR_DESCRIPTION = 'There was an error removing the user from the team';
export const VALIDATION_ERROR_TITLE = "Thereâ€™s a problem";
