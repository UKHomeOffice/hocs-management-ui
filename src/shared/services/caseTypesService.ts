import axios from 'axios';
import CaseType from '../models/caseType';
import Item from '../models/item';

const sortByLabel: ((a: Item, b: Item) => number) = (a, b) => {
    const labelA = a.label.toLocaleLowerCase();
    const labelB = b.label.toLocaleLowerCase();
    let comparison = 0;
    if (labelA > labelB) {
        comparison = 1;
    } else if (labelA < labelB) {
        comparison = -1;
    }
    return comparison;
};

export const getCaseTypes = () => new Promise<CaseType[]>((resolve, reject) =>
    axios.get('/api/case-types')
        .then((response) => {
            resolve(response.data.sort(sortByLabel));
        })
        .catch(error => reject(error))
);
