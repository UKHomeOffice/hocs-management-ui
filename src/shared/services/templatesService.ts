import axios from 'axios';

export const addTemplate = (template: FormData) => new Promise((resolve, reject) => axios
    .post('/api/templates', template)
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const deleteTemplate = (uuid: string) => new Promise((resolve, reject) => axios
    .delete(`/api/templates/${uuid}`)
    .then(() => resolve())
    .catch(reason => reject(reason))
);
