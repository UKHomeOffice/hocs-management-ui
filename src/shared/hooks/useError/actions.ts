import { FormError } from '../../models/formError';

export type AddFormError = {
    payload: FormError;
    type: 'AddFormError';
};

export type ClearError = {
    type: 'ClearError'
};

export type SetTitle = {
    payload: string;
    type: 'SetTitle';
};

export type SetDescription = {
    payload: string;
    type: 'SetDescription';
};

export type Action =
    AddFormError |
    ClearError |
    SetDescription |
    SetTitle;
