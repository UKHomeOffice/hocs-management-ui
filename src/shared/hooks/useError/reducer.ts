import { Action } from './actions';
import PageError from '../../models/pageError';

export const reducer = (state: PageError, action: Action): PageError => {
    switch (action.type) {
        case 'AddFormError':
            return {
                error: { ...state.error, formErrors: [...state.error ? state.error.formErrors || [] : [], action.payload] }
            };
        case 'ClearError':
            return { error: undefined };
        case 'SetDescription':
            return { error: { ...state.error, description: action.payload } };
        case 'SetTitle':
            return { error: { ...state.error, title: action.payload } };
    }
    return state;
};
