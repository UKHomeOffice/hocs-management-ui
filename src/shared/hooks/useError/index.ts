import React, { Reducer } from 'react';
import PageError from '../../models/pageError';
import { Action } from './actions';
import { reducer } from './reducer';
import { FormError } from '../../models/formError';
import ErrorMessage from 'shared/models/errorMessage';

const useError = (formErrorDescription?: string, formErrorTitle?: string)
    : [
        PageError,
        (formError: FormError) => void,
        () => void,
        (error: ErrorMessage) => void
    ] => {
    const [pageError, dispatch] = React.useReducer<Reducer<PageError, Action>>(reducer, {});

    const addFormError = (formError: FormError) => {
        (formErrorDescription || formErrorTitle) && setMessage({ title: formErrorTitle || '', description: formErrorDescription || '' });
        dispatch({ type: 'AddFormError', payload: formError });
    };
    const clearErrors = () => dispatch({ type: 'ClearError' });
    const setMessage = (errorMessage: ErrorMessage) => {
        dispatch({ type: 'SetDescription', payload: errorMessage.description });
        dispatch({ type: 'SetTitle', payload: errorMessage.title });
    };
    return [pageError, addFormError, clearErrors, setMessage];
};

export default useError;
