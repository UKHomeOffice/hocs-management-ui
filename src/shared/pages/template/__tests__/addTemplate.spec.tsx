import React from 'react';
import { match, MemoryRouter } from 'react-router-dom';
import { createBrowserHistory, History, Location } from 'history';
import { act, render, RenderResult, wait, fireEvent, waitForElement } from '@testing-library/react';
import AddTemplate from '../addTemplate';
import * as TemplatesService from '../../../services/templatesService';
import { GENERAL_ERROR_TITLE, ADD_TEMPLATE_ERROR_DESCRIPTION, LOAD_CASE_TYPES_DESCRIPTION } from '../../../models/constants';
import Template from '../../../models/template';
import * as useError from '../../../hooks/useError';
import * as CaseTypesService from '../../../services/caseTypesService';
import { createMockFile } from '../../../../../test/createMockFile';

jest.mock('../../../services/caseTypesService', () => ({
    __esModule: true,
    getCaseTypes: () => Promise.resolve([{
        label: '__caseType1__',
        value: '__CaseTypeId1__'
    }, {
        label: '__topic2__',
        value: '__topicId2__'
    }])
}));

const getTopicsSpy = jest.spyOn(CaseTypesService, 'getCaseTypes');

let match: match<any>;
let history: History<any>;
let location: Location;
let mockTemplate: Template;
let wrapper: RenderResult;

const useReducerSpy = jest.spyOn(React, 'useReducer');
const reducerDispatch = jest.fn();
const useErrorSpy = jest.spyOn(useError, 'default');
const addFormErrorSpy = jest.fn();
const clearErrorsSpy = jest.fn();
const setMessageSpy = jest.fn();

const renderComponent = () => render(
    <MemoryRouter>
        <AddTemplate history={history} location={location} match={match}></AddTemplate>
    </MemoryRouter>
);

jest.spyOn(TemplatesService, 'addTemplate').mockImplementation(() => Promise.resolve());

beforeEach(() => {
    history = createBrowserHistory();
    match = {
        isExact: true,
        params: { teamId: '__teamId__' },
        path: '',
        url: ''
    };

    location = {
        hash: '',
        key: '',
        pathname: '',
        search: '',
        state: {}
    };
    mockTemplate = {
        caseType: undefined,
        files: undefined
    };
    useReducerSpy.mockImplementation(() => [mockTemplate, reducerDispatch]);
    useErrorSpy.mockImplementation(() => [{}, addFormErrorSpy, clearErrorsSpy, setMessageSpy]);
    history.push = jest.fn();
    reducerDispatch.mockReset();
    addFormErrorSpy.mockReset();
    clearErrorsSpy.mockReset();
    setMessageSpy.mockReset();
    act(() => {
        wrapper = renderComponent();
    });
});

describe('when the addTemplate component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(1);

        await wait(() => {
            expect(wrapper.container).toMatchSnapshot();
        });
    });

    it('should display an error if the call to retrieve the case types fail', async () => {
        expect.assertions(1);
        getTopicsSpy.mockImplementation(() => Promise.reject('error'));

        act(() => {
            wrapper = renderComponent();
        });

        await wait(() => {
            expect(setMessageSpy).toBeCalledWith({ title: GENERAL_ERROR_TITLE, description: LOAD_CASE_TYPES_DESCRIPTION });
        });

    });
});

describe('when the submit button is clicked', () => {
    describe('and the data is filled in', () => {

        beforeEach(async () => {
            mockTemplate.files = [createMockFile()];
            mockTemplate.caseType = {
                label: '__caseTypeLabel__',
                value: '__caseTypeValue__'
            };
            const submitButton = await waitForElement(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        describe('and the service call is successful', () => {
            it('should redirect to the home page', async () => {
                expect.assertions(1);

                await wait(() => {
                    expect(history.push).toHaveBeenCalledWith('/', { successMessage: 'The template was created successfully' });
                });
            });
            it('should clear any previous errors', async () => {
                expect.assertions(1);

                await wait(() => {
                    expect(clearErrorsSpy).toHaveBeenCalled();
                });
            });
        });

        describe('and the service call fails', () => {
            beforeAll(() => {
                jest.spyOn(TemplatesService, 'addTemplate').mockImplementationOnce(() => Promise.reject({ response: { status: 500 } }));
            });

            it('should set the error state', () => {
                expect(setMessageSpy).toHaveBeenCalledWith({ description: ADD_TEMPLATE_ERROR_DESCRIPTION, title: GENERAL_ERROR_TITLE });
            });

            it('should clear any previous errors', () => {
                expect(clearErrorsSpy).toHaveBeenCalled();
            });
        });
    });
    describe('and the data is not filled in', () => {
        beforeEach(async () => {
            const submitButton = await waitForElement(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        it('should call the begin submit action', () => {
            expect(clearErrorsSpy).toHaveBeenCalled();
        });

        it('should set the error state', () => {
            expect(addFormErrorSpy).toHaveBeenNthCalledWith(1, { key: 'files', value: 'The Template is required' });
            expect(addFormErrorSpy).toHaveBeenNthCalledWith(2, { key: 'caseType', value: 'The Case Types is required' });
        });
    });
});
