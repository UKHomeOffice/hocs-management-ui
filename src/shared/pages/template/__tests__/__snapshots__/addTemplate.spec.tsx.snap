// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the addTemplate component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <a
        class="govuk-back-link"
        href="/case-type/__type__"
      >
        Back
      </a>
      <h1
        class="govuk-heading-xl"
      >
        Add a Template
      </h1>
      <h2
        class="govuk-heading-l"
      >
        Case Type: undefined
      </h2>
    </div>
  </div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-one-half-from-desktop"
    >
      <form
        action="/api/template"
        method="POST"
      >
        <input
          name="_csrf"
          type="hidden"
          value=""
        />
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label"
            for="files"
            id="files-label"
          >
            <label
              class="govuk-label govuk-label--s"
              for="files"
              id="files-label"
            >
              Add document
            </label>
          </label>
          <input
            class="govuk-file-upload"
            id="files"
            name="files"
            type="file"
          />
        </div>
        <input
          class="govuk-button"
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </div>
</div>
`;
